stl/inc/flat_map

```diff
@@ -0,0 +1,616 @@
+        else{
+            // Need to insert
+            auto _Index = _STD distance(begin(), _It);
+            _Insert_exact(_It, _STD move(_Val));
+            // Note: Consider _It invalidated by the insert operation
+            return _STD make_pair(begin() + _Index, true);
+        }
+    }
+
+    template <class... _Args_t>
+    iterator emplace_hint(const_iterator _Position, _Args_t&&... _Args) requires _STD is_constructible_v<value_type, _Args_t...> {
+        value_type _Val(_STD forward<_Args_t>(_Args)...);
+
+        const const_iterator _Begin = cbegin();
+        const const_iterator _End = cend();
+
+        // Adjust _Position
+        if(_Position == _End || !_Key_compare(*(_Position._Key_it), _Val.first)) {
+            // _Val should be inserted at or before _Position
+            if(_Position == _Begin || !_Key_compare(_Val.first, *((_Position - 1)._Key_it))){
+                // _Val should be inserted at or after _Position
+                // The hint _Position is accurate
+            }
+            else{
+                auto _Key_Position = _STD upper_bound(_Data.keys.cbegin(), _Position._Key_it, _Val.first, _Key_compare);
+                auto _Distance = _STD distance(_Data.keys.cbegin(), _Key_Position);
+                _Position = _Begin + _Distance;
+            }
+        }
+        else{
+            auto _Key_Position = _STD lower_bound(_Position._Key_it + 1, _Data.keys.cend(), _Val.first, _Key_compare);
+            auto _Distance = _STD distance(_Data.keys.cbegin(), _Key_Position);
+            _Position = _Begin + _Distance;
+        }
+
+        bool _Should_insert = false;
+        if(_Position == _End){
+            _Should_insert = true;
+        }
+        else if(_Key_equal(*(_Position._Key_it), _Val.first)){
+            _Should_insert = false;
+        }
+        else{
+            _Should_insert = true;
+        }
+
+        auto _Distance = _STD distance(_Begin, _Position);
+        if(_Should_insert){
+            _Insert_exact(_Position, _STD move(_Val));
+        }
+        // Consider _Position invalidated by the insert operation
+        return begin() + _Distance;
+    }
+
+    template <class _V>
+    _STD pair<iterator, bool> insert(_V&& _X) requires (same_as<_STD remove_cvref_t<_V&&>, value_type> || constructible_from<value_type, _V&&>) {
+        return emplace(_STD forward<_V>(_X));
+    }
+
+    template <class _V>
+    iterator insert(const_iterator _Position, _V&& _X) requires (same_as<_STD remove_cvref_t<_V&&>, value_type> || constructible_from<value_type, _V&&>) {
+        return emplace_hint(_Position, _STD forward<_V>(_X));
+    }
+
+    template <class _K, class... _Args_t>
+    _STD pair<iterator, bool> try_emplace(_K&& _Key, _Args_t&&... _Args)
+        requires constructible_from<mapped_type, _Args_t&&...>
+        && (same_as<_STD remove_cvref_t<_K&&>, key_type>
+            || (constructible_from<key_type, _K&&>
+                && _Compare_is_transparent<key_compare>

```

Comment:
Should implement a separate function for transparent comparators to avoid constructing Key before the actual insertion happens.